src/urbit.ts

```diff
@@ -0,0 +1,468 @@
+// import 'eventsource';
+import fetch from 'cross-fetch';
+import { isBrowser, isNode } from 'browser-or-node';
+
+import { AuthenticationInterface, SubscriptionInterface, CustomEventHandler, PokeInterface, SubscriptionRequestInterface, headers, UrbitInterface } from './index.d';
+import UrbitApp from './app/base';
+import { uncamelize } from './utils';
+
+
+// view-source:http://localhost/~landscape/js/channel.js
+// https://github.com/liam-fitzgerald/urbit-airlock-ts/blob/master/src/channel.ts
+
+/**
+ * A class for interacting with an urbit ship, given its URL and code
+ */
+class Urbit implements UrbitInterface {
+  /**
+   * Code is the deterministic password
+   */
+  code: string;
+
+  /**
+   * URL is the location of the instance.
+   */
+  url: string;
+
+  /**
+   * UID will be used for the channel: The current unix time plus a random hex string
+   */
+  uid: string;
+
+  /**
+   * Last Event ID is an auto-updated index of which events have been sent over this channel
+   */
+  lastEventId: number = 0;
+
+  lastAcknowledgedEventId: number = 0;
+
+  /**
+   * SSE Client is null for now; we don't want to start polling until it the channel exists
+   */
+  sseClient: EventSource | null = null;
+
+  /**
+   * Cookie gets set when we log in.
+   */
+  cookie?: string | null;
+
+  /**
+   * A registry of requestId to successFunc/failureFunc
+   * 
+   * These functions are registered during a +poke and are executed
+   * in the TODO onServerEvent()/onServerError() callbacks. Only one of
+   * the functions will be called, and the outstanding poke will be
+   * removed after calling the success or failure function.
+   */
+
+  outstandingPokes: Map<number, PokeInterface>;
+
+  /**
+   * A registry of requestId to subscription functions.
+   * 
+   * These functions are registered during a +subscribe and are
+   * executed in the onServerEvent()/onServerError() callbacks. The
+   * event function will be called whenever a new piece of data on this
+   * subscription is available, which may be 0, 1, or many times. The
+   * disconnect function may be called exactly once.
+   */
+
+  outstandingSubscriptions: Map<number, SubscriptionInterface>;
+
+  /**
+   * Ship can be set, in which case we can do some magic stuff like send chats
+   */
+  ship?: string | null;
+
+  /**
+   * If verbose, logs output eagerly.
+   */
+  verbose?: boolean;
+
+  /**
+   * All registered apps, keyed by name
+   */
+  static apps: Map<string, typeof UrbitApp> = new Map();
+
+  /** This is basic interpolation to get the channel URL of an instantiated Urbit connection. */
+  get channelUrl(): string {
+    return `${this.url}/~/channel/${this.uid}`;
+  }
+
+  /**
+   * Constructs a new Urbit connection.
+   *
+   * @param url  The URL (with protocol and port) of the ship to be accessed
+   * @param code The access code for the ship at that address
+   */
+  constructor(url: string, code: string) {
```

Comment:
```typescript
constructor(public url: string, public code: string)
```
Will save you have to assign them in the constructor and will also declare them as fields automatically.

Reply:
Neat